### PR TITLE
Feat/588 timelocks

### PR DESCRIPTION
**Pull request by Ramarti**
Adds timelocked action execution in accordance to [issue #558 ](https://github.com/balancer-labs/balancer-v2-monorepo/issues/588)

- Timelocked (delayed) actions are defined by using the method `setActionDelay` in `VaultAuthorizer`
- `setActionDelay` is in itself a delayed action, initially set with a delay of 1h
- Only a special kind of contract, `DelayedCall` can `trigger` a delayed action
- `DelayedCall`s are deployed calling `VaultAuthorizer.deployDelayedCall` by an account with permission granted for the `actionId` that would be triggered by `DelayedCall`
- `DelayedCall`s `trigger` can be permissionless or restricted to accounts with global permission granted for the correspondent `actionId`. This option is set when deploying them.
- `DelayedCall`s cancellation  is restricted to accounts with global permission granted for the correspondent `actionId`

### CODE REVIEW

**Comment by nventuro** [comment]
(I force-pushed a revised history where only the new commits are added, as squash-merging doesn't play nice with using branchs this way, particularly when merging back to the feature branch)

**Review by TomAFrench — COMMENTED** [review]
Hey Raúl :wave:. Was having a quick snoop through this PR and I have a question about the permissioning for delayed actions.

> `DelayedCalls` are deployed calling `VaultAuthorizer.deployDelayedCall` by an account with global permission granted for the `actionId` that would be triggered by `DelayedCall`

Considering we know the value of `where` for the `DelayedCall` to be deployed why would we not allow users for which `AccessControl.hasRole(actionId, account, where) == true` queue up this action?

File: pkg/vault/contracts/DelayedCall.sol [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    IDelayProvider private _delayProvider;
+    IAuthorizer private _authorizer;
+    bool public triggered = false;
+    uint256 public start;
+    address public where;
+    bytes public data;
+    bytes32 public actionId;
+    uint256 public value;
+    bool public immutable isTriggerPermissioned;
+    bool public cancelled;
```
[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/vault/contracts/DelayedCall.sol:30]: This is true for most of them, but I believe `immutable` is currently only supported for value types, so it won't work with arrays.

[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/vault/contracts/DelayedCall.sol:30]: True, maybe not `data` but I was just picking a field at random. Just as important to make sure that `where` can't change, etc.

**Comment by Ramarti** [comment]
> Hey Raúl wave. Was having a quick snoop through this PR and I have a question about the permissioning for delayed actions.
> 
> > `DelayedCalls` are deployed calling `VaultAuthorizer.deployDelayedCall` by an account with global permission granted for the `actionId` that would be triggered by `DelayedCall`
> 
> Considering we know the value of `where` for the `DelayedCall` to be deployed why would we not allow users for which `AccessControl.hasRole(actionId, account, where) == true` queue up this action?

Hi Tom, actually I just described it incorrectly, the method does actually allow for specific contracts:

```
    function deployDelayedCall(
        bytes32 actionId,
        address where,
        uint256 value,
        bytes calldata data,
        bool permissionedTrigger
    ) external returns (address) {
        require(AccessControl.hasRole(actionId, msg.sender, where), "Invalid permission");
        uint256 delay = _actionDelays[actionId];
        require(delay > 0, "Not a delayed action");
        DelayedCall delayedCall = new DelayedCall(data, where, value, this, this, permissionedTrigger, actionId);
        _delayedCalls[actionId].add(address(delayedCall));
        emit DelayedCallScheduled(actionId, address(delayedCall), where, value, data, delay);
        return address(delayedCall);
    }
    
   ```
   
   I'm going to update the description

File: pkg/vault/contracts/DelayedCall.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ramarti) [reply pkg/vault/contracts/DelayedCall.sol]: Good one, pushed a fix

**Comment by nventuro** [comment]
@facuspagnuolo is basing his work on the authorizer revamp milestone on this and other unmerged PRs, closing as this will be superceded by those.

**Pull request closed without merging**